File: architecture-skills/design-patterns/examples/strategy/strategy.java
// Strategy pattern in Java.
//
// The classic home of the pattern (java.util.Comparator is a strategy).
// Since Java 8 a single-method strategy interface is a functional
// interface, so lambdas replace one-off concrete classes; a named class
// is still useful when the strategy carries state (CarrierRate).

import java.util.LinkedHashMap;
import java.util.Map;

public class strategy {

    record Order(double weightKg, double distanceKm) {}

    @FunctionalInterface
    interface ShippingStrategy {
        double cost(Order order);
    }

    /** A stateful strategy: the negotiated rates are its state. */
    static final class CarrierRate implements ShippingStrategy {
        private final double base;
        private final double perKg;

        CarrierRate(double base, double perKg) {
            this.base = base;
            this.perKg = perKg;
        }

        @Override
        public double cost(Order order) {
            return base + perKg * order.weightKg();
        }
    }

    /** Context: delegates the pricing decision to the current strategy. */
    static final class Checkout {
        private ShippingStrategy strategy;

        Checkout(ShippingStrategy strategy) {
            this.strategy = strategy;
        }

        void setStrategy(ShippingStrategy strategy) {
            this.strategy = strategy;
        }

        double shippingFor(Order order) {
            return strategy.cost(order);
        }
    }

    public static void main(String[] args) {
        Order order = new Order(2.4, 120);

        Map<String, ShippingStrategy> strategies = new LinkedHashMap<>();
        strategies.put("flat", o -> 5.00);
        strategies.put("weight", o -> 4.00 + 1.50 * o.weightKg());
        strategies.put("distance", o -> 2.00 + 0.05 * o.distanceKm());
        strategies.put("carrier", new CarrierRate(3.00, 2.00));

        Checkout checkout = new Checkout(strategies.get("flat"));
        strategies.forEach((name, s) -> {
            checkout.setStrategy(s);
            System.out.printf("%8s: $%.2f%n", name, checkout.shippingFor(order));
        });
    }
}

File: architecture-skills/design-patterns/examples/strategy/strategy.cs
// Strategy pattern in C#.
//
// C# has both forms: Func<,> delegates/lambdas for stateless strategies
// (Comparison<T> and LINQ selectors are strategies in disguise), and a
// class when the strategy carries state (CarrierRate). The context
// accepts a delegate, so both plug in the same way.

using System;

public sealed record Order(double WeightKg, double DistanceKm);

// A stateful strategy: the negotiated rates are its state.
public sealed class CarrierRate
{
    private readonly double _base;
    private readonly double _perKg;

    public CarrierRate(double baseCost, double perKg)
    {
        _base = baseCost;
        _perKg = perKg;
    }

    public double Cost(Order order) => _base + _perKg * order.WeightKg;
}

// Context: delegates the pricing decision to the current strategy.
public sealed class Checkout
{
    private Func<Order, double> _strategy;

    public Checkout(Func<Order, double> strategy) => _strategy = strategy;

    public void SetStrategy(Func<Order, double> strategy) => _strategy = strategy;

    public double ShippingFor(Order order) => _strategy(order);
}

public static class Program
{
    public static void Main()
    {
        var order = new Order(2.4, 120);

        var strategies = new (string Name, Func<Order, double> Strategy)[]
        {
            ("flat", _ => 5.00),
            ("weight", o => 4.00 + 1.50 * o.WeightKg),
            ("distance", o => 2.00 + 0.05 * o.DistanceKm),
            ("carrier", new CarrierRate(3.00, 2.00).Cost),
        };

        var checkout = new Checkout(strategies[0].Strategy);
        foreach (var (name, s) in strategies)
        {
            checkout.SetStrategy(s);
            Console.WriteLine($"{name,8}: ${checkout.ShippingFor(order):F2}");
        }
    }
}
